### PR TITLE
Add new health check to service

### PR DESCRIPTION
**Pull request by nshumoogum**
### What

Add new health endpoint with the following checkers:

* elasticsearch
* hierarchy api
* kafka consumer
* kafka producer
* kafka error producer (for reporter)

Re-worked entry point to service via main and renamed service package to event and moved graceful shutdown logic to run functionality to be consistent with other applications.

Also includes:
* Use new go dependency manage, modules instead of vendor
* Use latest libraries, dp-kafka, dp-rchttp, dp-api-clients-go, log.go, dp-elasticsearch

### How to review

* Check unit tests pass
* Run an integration test
* Check code changes make sense

### Who can review

Anyone

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -8,21 +8,21 @@ import (
...
-	SignElasticsearchRequests bool          `envconfig:"SIGN_ELASTICSEARCH_REQUESTS"`
+	BindAddr                   string        `envconfig:"BIND_ADDR"`
+	Brokers                    []string      `envconfig:"KAFKA_ADDR"`
+	ConsumerGroup              string        `envconfig:"CONSUMER_GROUP"`
+	ConsumerTopic              string        `envconfig:"HIERARCHY_BUILT_TOPIC"`
+	ElasticSearchAPIURL        string        `envconfig:"ELASTIC_SEARCH_URL"`
+	EventReporterTopic         string        `envconfig:"EVENT_REPORTER_TOPIC"`
+	GracefulShutdownTimeout    time.Duration `envconfig:"GRACEFUL_SHUTDOWN_TIMEOUT"`
+	HealthCheckInterval        time.Duration `envconfig:"HEALTHCHECK_INTERVAL"`
+	HealthCheckCriticalTimeout time.Duration `envconfig:"HEALTHCHECK_CRITICAL_TIMEOUT"`
```
Comment (CarlHembrough) [inline config/config.go:19]: This also needs to be updated in the readme

File: event/handler.go [hunk 2]
```diff
@@ -85,7 +91,7 @@ func (svc *Service) handleMessage(ctx context.Context, message kafka.Message) (s
 		codeID := child.Links["code"].ID
 
 		if err = apis.addChildrenToSearchIndex(ctx, instanceID, dimension, codeID); err != nil {
-			log.Error(err, log.Data{"instance_id": instanceID, "dimension": dimension, "code_id": codeID})
+			log.Event(ctx, "faile to add children dimension options", log.ERROR, log.Error(err), log.Data{"instance_id": instanceID, "dimension": dimension, "code_id": codeID})
```
Comment (CarlHembrough) [inline event/handler.go]: typo "faile"

File: go.mod [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+go 1.13
+
+require (
+	github.com/ONSdigital/dp-api-clients-go v1.9.0
+	github.com/ONSdigital/dp-elasticsearch v1.1.0
+	github.com/ONSdigital/dp-healthcheck v1.0.2
+	github.com/ONSdigital/dp-hierarchy-api v0.0.0-20200309110613-76f2517bcbed
+	github.com/ONSdigital/dp-import v0.0.0-20180202121531-d3cc28e452c3
+	github.com/ONSdigital/dp-kafka v1.1.5
+	github.com/ONSdigital/dp-rchttp v0.0.0-20200114090501-463a529590e8
```
Comment (CarlHembrough) [inline go.mod]: Can this be replaced with a release version?

File: main.go [hunk 4]
```diff
@@ -2,99 +2,265 @@ package main
...
+		if err != nil {
+			log.Event(ctx, "failed to shutdown http server", log.ERROR, log.Error(err))
+			hasShutdownError = true
+		}
+		log.Event(shutdownContext, "stopped http server", log.INFO)
+
+		// If kafka consumer exists, stop listening to it. (Will close later)
+		if serviceList.Consumer {
+			log.Event(shutdownContext, "stopping kafka consumer listener", log.INFO)
+			syncConsumerGroup.StopListeningToConsumer(shutdownContext)
```
Comment (CarlHembrough) [inline main.go]: Should we be handling the error return value here?

File: main.go [hunk 5]
```diff
@@ -2,99 +2,265 @@ package main
...
+		if serviceList.SearchBuiltProducer {
+			log.Event(shutdownContext, "closing search builder kafka producer", log.INFO, log.Data{"topic": cfg.ProducerTopic})
+			searchBuiltProducer.Close(shutdownContext)
+			log.Event(shutdownContext, "closed search builder kafka producer", log.INFO, log.Data{"topic": cfg.ProducerTopic})
+		}
+
+		// If search builder error kafka producer exists, close it
+		if serviceList.SearchBuilderErrProducer {
+			log.Event(shutdownContext, "closing search builder error kafka producer", log.INFO, log.Data{"topic": cfg.EventReporterTopic})
+			searchBuilderErrProducer.Close(shutdownContext)
```
Comment (CarlHembrough) [inline main.go]: Should we be handling the error return value here?

File: main.go [hunk 6]
```diff
@@ -2,99 +2,265 @@ package main
...
+		if serviceList.Consumer {
+			log.Event(shutdownContext, "stopping kafka consumer listener", log.INFO)
+			syncConsumerGroup.StopListeningToConsumer(shutdownContext)
+			log.Event(shutdownContext, "stopped kafka consumer listener", log.INFO)
+		}
+
+		// If search built kafka producer exists, close it
+		if serviceList.SearchBuiltProducer {
+			log.Event(shutdownContext, "closing search builder kafka producer", log.INFO, log.Data{"topic": cfg.ProducerTopic})
+			searchBuiltProducer.Close(shutdownContext)
```
Comment (CarlHembrough) [inline main.go]: Should we be handling the error return value here?

File: main.go [hunk 7]
```diff
@@ -2,99 +2,265 @@ package main
...
+		// Close consumer loop
+		if err = consumer.Close(shutdownContext); err != nil {
+			log.Event(ctx, "failed to close consumer loop", log.ERROR, log.Error(err))
+			hasShutdownError = true
+		}
+
+		// If kafka consumer exists, close it
+		if serviceList.Consumer {
+			log.Event(shutdownContext, "closing kafka consumer", log.INFO, log.Data{"consumer": "SyncConsumerGroup"})
+			syncConsumerGroup.Close(shutdownContext)
```
Comment (CarlHembrough) [inline main.go]: Should we be handling the error return value here?

File: main.go [hunk 8]
```diff
@@ -2,99 +2,265 @@ package main
...
+	hasErrors := false
+
+	if err = hc.AddCheck("Kafka Consumer", kafkaConsumer.Checker); err != nil {
+		hasErrors = true
+		log.Event(ctx, "error adding check for kafka consumer", log.ERROR, log.Error(err))
+	}
+
+	if err = hc.AddCheck("Kafka Search Built Producer", searchBuiltProducer.Checker); err != nil {
+		hasErrors = true
+		log.Event(ctx, "error adding check for kafka import producer", log.ERROR, log.Error(err))
```
Comment (CarlHembrough) [inline main.go]: "error adding check for kafka search built producer"?

**Review by CarlHembrough — CHANGES_REQUESTED** [review]
Good work - just a few minor points to consider

File: go.mod (hunk 3 above)
Comment (nshumoogum) [reply go.mod]: There is currently no release for this library and James was updating the library with headers - https://github.com/ONSdigital/dp-rchttp/pull/9 - not sure whether we should be creating release 1 or not.

Comment (nshumoogum) [reply go.mod]: Actually I will create the first release and update the dependency here

**Review by CarlHembrough — APPROVED** [review]

**Pull request merged**